### PR TITLE
Sentry - Support browser

### PR DESCRIPTION
**Pull request by lejard-h**
- [x] Refactor to support browser
  + `SentryClient` from `package:sentry/sentry.dart` with conditional import
  + `SentryBrowserClient` for web from `package:sentry/browser_client.dart`
  + `SentryIOClient` for VM and Flutter from `package:sentry/io_client.dart`
- [x] Write test for browser
- [x] Working Angular Sentry packages => https://github.com/leftyio/angular_sentry

Main implementation difference are:
- gzip compressing not available on browser
- javascript stacktrace need to be prefix with `window.location.origin` to be resolve by sentry

### CODE REVIEW

**Comment by yjbanov** [comment]
Thanks for the PR! A couple of things before we can move forward. This PR needs two reviews, one to cover the dart-for-web side of things and one to cover the Flutter/VM parts. We have a few people on the Flutter team (including myself) who can provide the latter, but not the former. Therefore we need to find someone on our team who would be willing to own the Web functionality for flutter/sentry.

@kevmoo, @matanlurey, can you help with this?

If you need to move fast and can't wait for these organizational/ownership issues to be resolved, one option is to fork this library (Go open source!).

**Comment by lejard-h** [comment]
I already have a fork that we use in production and that is used by angular_sentry for a few month now.

I don't need to move fast, the coming release of dart 2 and angular 5 seems to be a good opportunity to build a strong library for sentry.

**Comment by matanlurey** [comment]
It is unlikely my team will be able to own this anytime soon @yjbanov

**Comment by lejard-h** [comment]
angular implementation => https://github.com/leftyio/angular_sentry/pull/1/

**Comment by Hixie** [comment]
@yjbanov Are you able to continue to shepherd this through?

**Comment by yjbanov** [comment]
Unfortunately, no.

**Comment by Hixie** [comment]
@lejard-h We've moved this to the `packages` repo. I'm not sure we have anyone who would be able to review the change, still, but in the meantime, would you be willing to resubmit the PR to the other repo?

**Comment by listepo** [comment]
Any news? (Flutter for a web already presented)

**Comment by yjbanov** [comment]
> Any news? (Flutter for a web already presented)

Yes, this will have to align with the Flutter for web efforts, likely after we merge the Web support into Flutter proper.

**Comment by Standaa** [comment]
Any updates as to when this PR will be accepted ?

**Comment by yjbanov** [comment]
As of September 10, 2019, Flutter for web merged into the mainline Flutter SDK, so we can certainly reopen this discussion.

**Comment by lejard-h** [comment]
I fixed the conflicts.

Test are passing but I still need to check my angular_sentry implementation and run additional tests on real env.

I also added conditional import support on `package:sentry/sentry.dart` so it should not be breaking change.

**Comment by jaumard** [comment]
I've added sentry without checking first if it's support web :( shame on me lool can't run on web anymore. 

But now I'm interested to see this merged ^^

**Comment by yjbanov** [comment]
I'm reviewing this PR right now in between other high-priority tasks that have been assigned to me, but it's big! :) So it will take some time.

File: tool/dart2_test.sh [hunk 1]
```diff
@@ -1,7 +0,0 @@
-#!/bin/sh
```
Comment (yjbanov) [inline tool/dart2_test.sh:1]: Nice clean-up! Thanks.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
 # package:sentry changelog
 
+## 2.4.0
```
Comment (yjbanov) [inline CHANGELOG.md]: How about making it a `3.0.0-dev0`, given how big and probably subtly breaking this change is?

File: lib/src/base.dart [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+import 'dart:convert';
+
+import 'package:http/http.dart';
+import 'package:meta/meta.dart';
+import 'package:usage/uuid/uuid.dart';
+
+import 'client_stub.dart'
+    // ignore: uri_does_not_exist
+    if (dart.library.html) 'browser.dart'
+    // ignore: uri_does_not_exist
```
Comment (yjbanov) [inline lib/src/base.dart]: Why are ignores necessary?

File: test/sentry_browser_test.dart [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+import 'package:test/test.dart';
+
+const String _testDsn = 'https://public:secret@sentry.example.com/1';
+const String _testDsnWithoutSecret = 'https://public@sentry.example.com/1';
+const String _testDsnWithPath =
+    'https://public:secret@sentry.example.com/path/1';
+const String _testDsnWithPort =
+    'https://public:secret@sentry.example.com:8888/1';
+void main() {
+  group('$SentryBrowserClient', () {
```
Comment (yjbanov) [inline test/sentry_browser_test.dart]: You can also do `group(SentryBrowserClient, ...`

File: test/sentry_browser_test.dart [hunk 5]
```diff
@@ -0,0 +1,425 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline test/sentry_browser_test.dart:1]: It looks like a lot of the test code is identical with the `io` version. Would it be possible to have shared test code configurable via a `typedef SentryClientFactory` that provides either `io` or `browser` implementation of the client??

**Review by yjbanov — COMMENTED** [review]
Do `browser_client.dart` and `io_client.dart` add much value on top of the configurable imports?

File: lib/src/stack_trace.dart [hunk 6]
```diff
@@ -70,3 +82,9 @@ String _absolutePathForCrashReport(Frame frame) {
 
   return '${frame.uri}';
 }
+
+class EmptyStacktraceException implements Exception {
```
[the remark this answers is not in the document]
Comment (lejard-h) [reply lib/src/stack_trace.dart]: Actually It was not used, It just removed it

File: test/sentry_browser_test.dart (hunk 5 above)
Comment (lejard-h) [reply test/sentry_browser_test.dart:1]: The conditional import made it really simple to do, I moved all common and configurable tests under the `test_utils.dart` file using a single function 
```dart
runTest({Codec<List<int>, List<int>> gzip, bool isWeb = false})
```

Also moved `Event` test under `event_test.dart` file.

**Comment by jaumard** [comment]
All good now @yjbanov ? :)

File: lib/sentry.dart [hunk 7]
```diff
@@ -1,653 +1,8 @@
-// Copyright 2017 The Chromium Authors. All rights reserved.
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline lib/sentry.dart]: Let's keep the original year.

File: lib/src/client_stub.dart [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import 'package:http/http.dart';
```
Comment (yjbanov) [inline lib/src/client_stub.dart:5]: This needs a license header.

**Comment by yjbanov** [comment]
Thanks for the ping @jaumard.

@lejard-h Is this ready for another pass? This looks almost there.

File: lib/src/client_stub.dart (hunk 8 above)
Comment (lejard-h) [reply lib/src/client_stub.dart:5]: should be good now

**Comment by yjbanov** [comment]
@lejard-h I can't thank you enough for your boundless patience! As I read through the history of this PR, I'm amazed at how things have changed since August 2018, when it was first filed. This PR survived tectonic shifts in Flutter and Dart.

lgtm

**Pull request merged**